### ISSUE TITLE
Make dxtbx format_kwargs pass along arbitrary parameters

### ISSUE THREAD

**Issue by phyy-nx** [cctbx/cctbx_project#99]
Currently two parameters are passed along to the format objects in the kwargs dictionary of their init functions, dynamic_shadowing and multi_panel.  These are defined in two locations:
1) A format phil blob in util/options.py
2) A try except block in util/options.py where they are read from the phil blob by the importer

If this phil blob could take arbitrary parameters, then format classes could take arbitrary parameters.  This would be super helpful for metadata poor data files and would allow less hardcoding of parameters in format objects.

**Comment by jmp1985** [comment]
We can certainly add other parameters here if there are other parameters you need in a format class.

**Comment by phyy-nx** [comment]
Sure, but then this phil blob becomes an arbitrary bucket of random
parameters.  Making it general maybe is better?

**Comment by jmp1985** [comment]
I think these parameters are for optional interpretation within the format class.  If we want to override meta data we can do so with the geometry {} phil parameters. I guess if it looks like we'll need lots more parameters like this then we could make it more general. What do you have in mind?

**Comment by phyy-nx** [comment]
Right, so the specific use case I'm considering is XTC data from LCLS.  A
putative XTC reader needs to know several parameters too numerous to list
here, but a good example is a detector address which identifies a detector
by name to pull out of an xtc stream which can have multiple detector
types.  We could use a whole phil blob to define all the parameters needed
to read the data, but that would clutter options.py unnecessarily.  Having
a generic way to pass information along to a format class seems important.

**Comment by jmp1985** [comment]
Hi Aaron,

So I don't have any experience with XTC streams but wouldn't it be possible to encode the knowledge of the XTC stream in the format class?

James

________________________________

**Comment by phyy-nx** [comment]
Continuing with the example of the detector address, no it's not.  With
hundreds of experiments at LCLS, the detectors have many different names.
Not too mention the other details I didn't go into, such as detz_offset
(different from detector distance) or the calibration directory location.

**Comment by jmp1985** [comment]
OK. So I suppose we don't want to put dozens of beamline specific parameters here. 

One way of doing this might be to have a parameter that points to a phil file containing format specific parameters like:

```
format {
  phil_file=format_specific_parameters.phil 
}
```

This would have the benefit of not adding lots of parameters here and would be easy to implement for passing the arguments. However, it's not very user friendly in that the parameters wouldn't be visible on the command line.

Another option would be to have a configuration file maybe not in phil but maybe json or something that does the same job.

Finally, we could have scoped parameters loaded from the format classes such that we might have:

format {
  SubFormat {
    SubSubFormat1 {
      parameter = something
    }
    SubSubFormat2 {
      another_parameter = something_else
    }
  }  

}

We could generate these from the format class. This would have the benefit of being flexible and having all the parameters on the command line in a familiar way. However, it might potentially lead to parameter duplication and I think that format classes would need to inherit parameters from their parents so the whole thing might get complicated if lots of format classes used this mechanism. In particular, if we generated from the format class for the multi_panel parameter, it would mean that every time you typed dials.import multi_panel=True, it would say "which parameter did you mean format1.multi_panel, format.multi_panel etc" and it would become a pain.

Can you think of another way to achieve this that might be better

**Comment by phyy-nx** [comment]
I do like inherited scopes.  Let us prototype something over here and get back to you.
